bee-network/src/service/service.rs

```diff
@@ -0,0 +1,485 @@
+            peer_id,
+            peer_addr,
+            conn_info,
+            gossip_in,
+            gossip_out,
+        } => {
+            let mut peerlist = peerlist.0.write().await;
+
+            // In case the peer doesn't exist yet, we create a `PeerInfo` for that peer on-the-fly.
+            if !peerlist.contains(&peer_id) {
+                let peer_info = PeerInfo {
+                    address: peer_addr,
+                    alias: alias!(peer_id).to_string(),
+                    relation: PeerRelation::Unknown,
+                };
+
+                peerlist
+                    .insert_peer(peer_id, peer_info.clone())
+                    .map_err(|(_, _, e)| e)?;
+
+                let _ = senders.events.send(Event::PeerAdded {
+                    peer_id,
+                    info: peer_info,
+                });
+            }
+
+            assert!(peerlist.contains(&peer_id));
+
+            // Unwrap:
+            // We made sure, that the peer exists.
+            let peer_info = peerlist.info(&peer_id).unwrap();
+
+            // We store a clone of the gossip send channel in order to send a shutdown signal.
+            let _ = peerlist.update_state(&peer_id, |state| state.set_connected(gossip_out.clone()));
+
+            info!(
+                "Established ({}) protocol with {} ({}).",
+                conn_info.origin,
+                peer_info.alias,
+                alias!(peer_id)
+            );
+
+            let _ = senders.events.send(Event::PeerConnected {
+                peer_id,
+                info: peer_info,
+                gossip_in,
+                gossip_out,
+            });
+        }
+    }
+
+    Ok(())
+}
+
+async fn add_peer(
+    peer_id: PeerId,
+    address: Multiaddr,
+    alias: String,
+    relation: PeerRelation,
+    senders: &Senders,
+    peerlist: &PeerList,
+) -> Result<(), PeerError> {
+    let peer_info = PeerInfo {
+        address,
+        alias,
+        relation,
+    };
+
+    let mut peerlist = peerlist.0.write().await;
+
+    // If the insert fails for some reason, we get the peer data back, so it can be reused.
+    match peerlist.insert_peer(peer_id, peer_info) {
+        Ok(()) => {
+            // NB: We could also make `insert` return the just inserted `PeerInfo`, but that would
+            // make for an unusual API.
+
+            // We just added it, so 'unwrap'ping is safe!
+            let info = peerlist.info(&peer_id).unwrap();
+
+            let _ = senders.events.send(Event::PeerAdded { peer_id, info });
+
+            Ok(())
+        }
+        Err((peer_id, peer_info, mut e)) => {
+            // NB: This fixes an edge case where an in fact known peer connects before being added by the
+            // manual peer manager, and hence, as unknown. In such a case we simply update to the correct
+            // info (address, alias, relation).
+
+            if matches!(e, PeerError::PeerIsAdded(_))
+            // && peer_info.relation.is_known()
```

Comment:
should this comment be deleted?

Reply:
it should :trollface:

Reply:
Deleted and added explanatory context to the method.